File: packages/ink-swift/Tests/InkGravesTests/EngineTests.swift
/// End-to-end engine behavior, ported from
/// packages/ink-graves/test/engine.test.ts.

import Foundation
import InkGraves
import Testing

@Suite struct EngineTests {
    let model = try! GravesModel(assets: Fixtures.assets)

    @Test func writesAnUnprimedLineAndTerminatesNaturally() throws {
        let offsets = try model.write("hello world", bias: 0.75, seed: 42)
        #expect(offsets.count > 100)
        #expect(offsets.count < 40 * "hello world".count)
        for offset in offsets {
            #expect(offset.dx.isFinite)
            #expect(offset.dy.isFinite)
        }
        // eos=1 marks the last point of a stroke; a legible line has several.
        let strokeCount = offsets.filter(\.eos).count
        #expect(strokeCount > 3)
    }

    @Test func isDeterministicPerSeedAndVariesAcrossSeeds() throws {
        let first = try model.write("hello", bias: 0.75, seed: 7)
        let second = try model.write("hello", bias: 0.75, seed: 7)
        let different = try model.write("hello", bias: 0.75, seed: 8)
        #expect(first == second)
        #expect(first != different)
    }

    @Test func writesWithStylePriming() throws {
        let offsets = try model.write("hello", bias: 0.75, style: 9, seed: 42)
        #expect(offsets.count > 40)
        #expect(offsets.allSatisfy { $0.dx.isFinite && $0.dy.isFinite })
    }

    @Test func rejectsUnknownStylesAndOverLongText() {
        #expect(throws: GravesError.self) {
            try model.writer("hi", style: 99)
        }
        #expect(throws: GravesError.self) {
            try model.writer(String(repeating: "x", count: 200))
        }
    }

    @Test func sustainsAtLeast125StepsPerSecond() throws {
        let writer = try model.writer(
            "the quick brown fox jumps over the lazy dog then keeps on going",
            bias: 0.75,
            seed: 1
        )
        for _ in 0 ..< 30 { _ = writer.step() } // warmup
        let clock = ContinuousClock()
        var steps = 0
        let elapsed = clock.measure {
            while steps < 600, writer.step() != nil { steps += 1 }
        }
        let stepsPerSecond = Double(steps) / max(elapsed.seconds, 1e-9)
        print("engine speed: \(Int(stepsPerSecond)) steps/sec over \(steps) steps")
        #expect(stepsPerSecond > 125)
    }
}

private extension Duration {
    var seconds: Double {
        let (seconds, attoseconds) = components
        return Double(seconds) + Double(attoseconds) / 1e18
    }
}

File: packages/ink-kotlin/ink-graves/src/test/kotlin/com/trylonghand/ink/graves/EngineTests.kt
/// End-to-end engine behavior, ported from
/// packages/ink-graves/test/engine.test.ts.

package com.trylonghand.ink.graves

import kotlin.math.max
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue

class EngineTests {
    private val model = GravesModel(Fixtures.assets)

    @Test
    fun writesAnUnprimedLineAndTerminatesNaturally() {
        val offsets = model.write("hello world", bias = 0.75, seed = 42u)
        assertTrue(offsets.size > 100)
        assertTrue(offsets.size < 40 * "hello world".length)
        for (offset in offsets) {
            assertTrue(offset.dx.isFinite())
            assertTrue(offset.dy.isFinite())
        }
        // eos=1 marks the last point of a stroke; a legible line has several.
        val strokeCount = offsets.count { it.eos }
        assertTrue(strokeCount > 3)
    }

    @Test
    fun isDeterministicPerSeedAndVariesAcrossSeeds() {
        val first = model.write("hello", bias = 0.75, seed = 7u)
        val second = model.write("hello", bias = 0.75, seed = 7u)
        val different = model.write("hello", bias = 0.75, seed = 8u)
        assertEquals(first, second)
        assertNotEquals(first, different)
    }

    @Test
    fun writesWithStylePriming() {
        val offsets = model.write("hello", bias = 0.75, style = 9, seed = 42u)
        assertTrue(offsets.size > 40)
        assertTrue(offsets.all { it.dx.isFinite() && it.dy.isFinite() })
    }

    @Test
    fun rejectsUnknownStylesAndOverLongText() {
        assertFailsWith<GravesError> {
            model.writer("hi", style = 99)
        }
        assertFailsWith<GravesError> {
            model.writer("x".repeat(200))
        }
    }

    /**
     * Real-time streaming needs 125 steps/sec (the web app reveals one
     * step per 8ms). The graves cell's matrix loops are plain Kotlin, but
     * the JIT-compiled saxpy inner loop clears the bar comfortably.
     */
    @Test
    fun sustainsAtLeast125StepsPerSecond() {
        val writer = model.writer(
            "the quick brown fox jumps over the lazy dog then keeps on going",
            bias = 0.75,
            seed = 1u,
        )
        repeat(30) { writer.step() } // warmup
        var steps = 0
        val start = System.nanoTime()
        while (steps < 600 && writer.step() != null) steps += 1
        val elapsed = (System.nanoTime() - start) / 1e9
        val stepsPerSecond = steps / max(elapsed, 1e-9)
        println("engine speed: ${stepsPerSecond.toInt()} steps/sec over $steps steps")
        assertTrue(stepsPerSecond > 125)
    }
}
